integration/integration.go

```diff
@@ -214,12 +220,9 @@ func QEMUTest(t *testing.T, o *Options) (*qemu.VM, func()) {
 
 	return vm, func() {
 		vm.Close()
-		dir := vm.Options.Devices[0].(qemu.ReadOnlyDirectory).Dir
-		if t.Failed() {
-			t.Log("keeping temp dir: ", dir)
-		} else if len(o.TmpDir) == 0 {
-			if err := os.RemoveAll(dir); err != nil {
-				t.Logf("failed to remove temporary directory %s: %v", dir, err)
+		if len(o.TmpDir) == 0 {
+			for _, dev := range vm.Options.Devices {
+				dev.Cleanup(t)
```

Comment:
I think instead of having Devices have a cleanup function, maybe QEMU() should just return a cleanup() func, rather than trying to get across the abstraction here.

Reply:
Not quite certain what you're suggesting. Move the function beginning on L221 into QEMU(), return it there, and pass it up to this function's caller? Or...?

Reply:
Actually came up with something different. Hey, I screwed it up, so I'll clean it up: #1295

Reply:
I really like having the cleanup functions tied to the Devices that are making the mess. IIUC you're trying to avoid that with 1295?

Perhaps the cleanup logic will never extend beyond the tempdir, but I am using the integration and qemu packages with code at work, and sooner or later I'll be implementing some custom Devices that might be a little harder to clean up after.